File: SoftwareEngineering/DesignPattern/AbstractFactory/typescript/AbstractFactory.ts
export interface IButton {
    paint(): void;
}

export interface IGUIFactory {
    createButton(): IButton;
}

export class WinFactory implements IGUIFactory {
    public createButton(): IButton {
        return new WinButton();
    }
}

export class OSXFactory implements IGUIFactory {
    public createButton(): IButton {
        return new OSXButton();
    }
}

export class WinButton implements IButton {
    public paint(): void {
        //Render a button in a Windows style
    }
}

export class OSXButton implements IButton {
    public paint(): void {
        //Render a button in a Mac OS X style
    }
}

const render = (appearance: string): IGUIFactory => {
    let factory: IGUIFactory;
    switch (appearance) {
        case "Win":
            factory = new WinFactory();
            break;
        case "OSX":
            factory = new OSXFactory();
            break;
        default:
            throw new Error('Not Implemented');
    }
    return factory;
};

const appearance = 'Win';
const factory: IGUIFactory = render(appearance);

const button = factory.createButton();
button.paint();

File: SoftwareEngineering/DesignPattern/AbstractFactory/csharp/AbstractFactory.cs
interface IButton
{
    void Paint();
}

interface IGUIFactory
{
    IButton CreateButton();
}

class WinFactory : IGUIFactory
{
    public IButton CreateButton()
    {
        return new WinButton();
    }
}

class OSXFactory : IGUIFactory
{
    public IButton CreateButton()
    {
        return new OSXButton();
    }
}

class WinButton : IButton
{
    public void Paint()
    {
        //Render a button in a Windows style
    }
}

class OSXButton : IButton
{
    public void Paint()
    {
        //Render a button in a Mac OS X style
    }
}

class Program
{
    static void Main()
    {
        var appearance = Settings.Appearance;

        IGUIFactory factory;
        switch (appearance)
        {
            case Appearance.Win:
                factory = new WinFactory();
                break;
            case Appearance.OSX:
                factory = new OSXFactory();
                break;
            default:
                throw new System.NotImplementedException();
        }

        var button = factory.CreateButton();
        button.Paint();
    }
}
